### DISCUSSION TITLE
How to add open telemetry traces

### DISCUSSION THREAD

**Discussion by RichardWright** [ReactiveX/rxjs#6615] [Q&A]
Has anyone ever made a suggestion on how to add spans/traces to an rxjs observable? Specifically on something like receiving data from a websocket?

**Comment by abarke** [comment]
You can create a new **serialize** and **deserialize** methods to intercept all messages before/after they are received/transmitted.

The functions can be set in the [WebSocketSubjectConfig](https://rxjs.dev/api/webSocket/WebSocketSubjectConfig#serializer)

**Reply by RichardWright** [reply]
@abarke great suggestion! is there a context object on each observable where you get the opening span to close it?

**Reply by abarke** [reply]
@RichardWright not sure I understand. Do you have a use case for what you want to achieve to give some more context?

**Reply by RichardWright** [reply]
Hi!

Basic example of tracing here - https://www.npmjs.com/package/@opentelemetry/tracing

I could run the below line inside a deserialiser 

`const span = opentelemetry.trace.getTracer('default').startSpan('foo');`


But the following, I'm not sure where I could store then retrieve the span from?

`// We must end the spans so they become available for exporting.
span.end();`

I could have a global store for them, but creating an identity for them to retrieve them is something I fall down on?

**Reply by abarke** [reply]
At what point in the lifecycle do you need to start and end the span? Is there an concrete example you could provide?

Does this apply to sending a particular request and expecting a response with a specific ID?

**Reply by RichardWright** [reply]
I would start when deserializing. I would end at some later point, ideally one where I control it, just before leaving redis. All other subscriptions inside rxjs would be seperate, nested spans

**Reply by abarke** [reply]
I suppose you would somehow need to define an array (of spans) of sufficient scope so that each span can be ended gracefully at some point. Sounds questionable though. I don't really know enough to give a good suggestion, hence wondering what the overall goal is.

**Reply by RichardWright** [reply]
Goal would to be able to measure the observables involved (spans in this case) between the arrival of a packet over a websocket to a final step where it is readied for publishing to a pubsub system. Ideally in a way which inline with rxjs practises and isn't obtrusive.

**Reply by RichardWright** [reply]
@abarke Another way of looking at this -

Is there a way of adding context to observable emitance that would allow state to be stored and therefore a span to be retrieved?

**Reply by abarke** [reply]
@RichardWright are you using the WebSocketSubject?

If the start of the span is done when deserializing and the end of the span is some time later in the Observable stream, then you can use the `tap` operator to trigger side-effects without modifying the data.

**Reply by RichardWright** [reply]
As in, use a closure?

This observable is returned and subsequent subscriptions wouldn't have access to it's scope.

**Reply by abarke** [reply]
If a observable is returned the you could simply create a new observable from it and do:

```javascript
const newObservable = observable.pipe(
  tap((val) => console.log(val) && endSpan()) 
)
return newObservable
```

I don't know the whole scope of what your code looks like so perhaps some code examples would clarify things a bit?

**Reply by RichardWright** [reply]
We usually combine them, eg - combineLatest([ob1, ob2, ob3]).pipe() and they are shared in between various pieces of code.

I think my example might not be helping.

Would wrapping the responses from the observables in an object with tracers etc on it be a better approach?

eg return { tracer: { spanId}, data: dataToReturn }

**Reply by abarke** [reply]
I would assume having a tracer span ID would make sense as you need to identify the individual responses. I'm no RxJS expert, just happened to find this topic interesting 😊

**Reply by RichardWright** [reply]
an id would allow for a central place to store the actual span/tracer. It could equally be the actual span.

I feel like i'm asking the wrong questions and this has already been solved in rxjs somewhere else.

**Reply by scvnc** [reply]
I was thinking about the `{ tracer: {spanId}, data: dataToReturn }` pattern too.  In typescript it could be handled like a

```ts
type Envelope<T> {
  data: T,
  _span: Span
}
```
Might be a way to build a toolset of RxJS operators that notice such an Envelope signature and create a child span off of it.

```ts
const {startSpan, finalizeSpan} =  createRxJsTracer()

someEvent$.pipe(
  startTrace('some-interesting-service-call'),
  // do some remote call
  switchMap((envelope) => ajax('/some/api', data: { id: envelope.data.id }),
  finalizeSpan, // handles on stream errors too
)

```
but still figuring it out.



let me know if you'd like to collaborate on this in further detail. I am curious to build/find some tooling/patterns to bolt in tracing/spans into a rxjs heavy SPA UI.

**Reply by RichardWright** [reply]
I'd love to work on this. I was mostly battling my current type system, would love to be able to explore the problem with someone similarly motivated.

**Reply by RichardWright** [reply]
@scvnc

**Reply by scvnc** [reply]
Hi @RichardWright 

Since I was last here, I've been experimenting working on an Rxjs OperatorFn that uses a javascript mixin behind the scenes. Allows a span to be created/completed/error-handled when something subscribes and data flows through.

 Everyones use case may be different, but so far this is working for me... gotta figure out nested spans.. E.g. that doHttpCall might also create a span.
 
 ```ts
    id$.pipe(
        withSpan('span-name', optionsObj, _id$ =>
            _id$.pipe(
                // Do work in this composition and it will be tracked in the context of this span
                switchMap((id) => doHttpCall(id)),
                map(transformResult)
            )
        ),
       someWorkOutsideOfSpan()
    );
 ```
